packages/core/core/src/ConfigLoader.js

```diff
@@ -0,0 +1,103 @@
+// @flow
+import ConfigResolver from './ConfigResolver';
+import {breakStatement} from 'babel-types';
+import {conditionalExpression} from '@babel/types';
+
+export default class ConfigLoader {
+  constructor(options) {
+    this.options = options;
+  }
+
+  load(configRequest) {
+    if (configRequest.configType === 'parcel') {
+      return this.loadParcelConfig(configRequest);
+    }
+
+    return this.loadThirdPartyConfig(configRequest);
+  }
+
+  async loadParcelConfig(configRequest) {
+    let {filePath} = configRequest;
+    let configResolver = new ConfigResolver();
+
+    // Resolve plugins from cwd when a config is passed programmatically
+    let config = this.options.config
+      ? await configResolver.create({
+          ...this.options.config,
+          resolveFrom: this.options.cwd
+        })
+      : await configResolver.resolve(filePath);
+    if (!config && this.options.defaultConfig) {
+      config = await configResolver.create({
+        ...this.options.defaultConfig,
+        resolveFrom: this.options.cwd
+      });
+    }
+
+    if (!config) {
+      throw new Error('Could not find a .parcelrc');
+    }
+
+    this.parcelConfig = config;
+
+    let devDeps = [];
+    switch (configRequest.meta.actionType) {
+      case 'transformer_request':
+        devDeps = config.getTransformerNames(filePath);
+        break;
+      case 'dependency':
+        devDeps = config.getResolverNames();
+        break;
+    }
+    let devDepRequests = devDeps.map(devDep => ({
+      moduleSpecifier: devDep,
+      resolveFrom: config.resolveFrom // TODO: resolveFrom should be nearest package boundary
+    }));
+
+    let invalidations = [
+      {
+        action: 'add',
+        pattern: '**/.parcelrc'
+      },
+      {
+        action: 'change',
+        pattern: config.filePath
+      },
+      {
+        action: 'unlink',
+        pattern: config.filePath
+      }
+    ];
+
+    let reliesOnLockFile = false;
+    for (let extendedFile of config.extendedFiles) {
+      // ? Does this work for Windows
+      if (extendedFile.includes('/node_modules/')) {
+        reliesOnLockFile = true;
+      }
+      invalidations.push({
+        action: 'change',
+        pattern: extendedFile
+      });
+    }
+
+    // These are only needed for invalidations and do not need to be included in the hash
+    // TODO: probably shouldn't get rid of them this way though
+    delete config.extendedFiles;
+    delete config.filePath;
+
+    if (reliesOnLockFile) {
```

Comment:
If any of the extended configs come from packages instead of relative files, then they should be resolved again when the lockfile changes because it may be a newer version. That means we need to invalidate the `config_request` when the lock file changes and rebuild the config. If there are no extended configs that come from packages then we don't need to invalidate.

Reply:
Should it rely on the `package.json` inside that node_module instead of the lockfile? Lockfiles aren't required, and that would have the same effect, no?

Reply:
That doesn't work very well for Yarn plug'n'play because the package.json at that location will never change. It also doesn't work very well with normal node module resolution if a different version of the same package shows up in a closer node_modules folder. Checking the lock file is a very simple way to check whether we should resolve again or not. If there is no lockfile, I'm thinking about adding some way to tell parcel that something should be invalidated on start up, so that might work here.